Intermittent test failure

CI has failed due to this error but it worked when restarted: 

```
error: test failed, to rerun pass `--lib`

failures:

---- nn::transformer::decoder::tests::test_autoregressive_norm_last stdout ----
thread 'nn::transformer::decoder::tests::test_autoregressive_norm_last' panicked at 'assertion failed: `(left == right)`
  left: `[0.23948555, -1.4476444, 1.7394568, -0.37974617, -0.7028924, -0.4129355, -1.9158951, 0.5557, 1.1982789, 0.40759498, 0.06550879, 0.6530054, 0.096770726, -0.04666269, 1.3337429, -1.6876903, -0.40740308, -0.12567851, -0.7417101, 0.33780476, 2.0305002, -0.5805786, -1.112808, 0.90371263, -0.42485318, -1.6029015, 0.774434, -0.54886544, 0.43462497, -0.18808673, -2.1350539, 0.20949733, 0.87836653, 0.41200003, 0.7943648, 1.3964221, -0.16162446, 0.8851795, 1.745951, -1.2020794, -0.7502541, 0.5759191, -1.9597254, 0.059071593, 0.9495946, 0.63463026, -0.80792814, 0.031265542, -0.8413961, -1.1042768, 0.6303067, -1.8922045, 0.44333157, -0.24039328, -1.1531141, 0.4702631, 1.3636328, 0.12676455, 1.1119264, 1.0851598, -0.17424995, 0.091177955, 0.52813435, -2.269342, 0.7240235, 1.2046725, -1.5445613, 0.43560714, 1.3055882, -0.08621346, 0.2435217, -0.4583588, 0.17207853, 0.71333534, 1.3131536, -1.8801376, -1.7852423, 0.68362814, -0.06631981, 1.0313472, -0.47530693, -0.68399745, 0.096292794, 0.88116866, 0.9159162, 0.30998585, 1.8870828, -1.6558123, 0.27701497, -0.91328967, -1.6643143, 0.77425784, -0.114014834, -0.4628347, 0.47438326, 0.17162472, 0.3830775, -1.7981151, 1.4499048, -1.6079453, 1.0426321, -0.4928484, -0.70687276, -0.18113491, 0.69756544, -0.13894956, 1.2459149, 0.10677127, -1.2047524, -0.7041038, 0.7481061, -0.77757, -0.9965105, -0.5658682, -0.24453977, -0.19981912, 2.489774, 0.44647074, -0.02859344, 1.0374068, 0.23086856, 0.18305853, 1.7442065, -0.14733128, -2.0698054, -0.04084435, -0.86746764, 1.4469825, 0.21824387, -0.50633633, -0.93084013, 0.7392653, -0.6499252, 0.6358838, 1.3024436, -0.9860946, -0.40290937, 0.5891721, -1.7343421, 0.30868047, 2.050654, -0.6654694, -0.49161807, 0.043524742]`,
 right: `[0.23948555, -1.4476444, 1.7394568, -0.37974617, -0.7028924, -0.4129355, -1.9158951, 0.555783, 1.1982789, 0.40759498, 0.06550879, 0.6530054, 0.09677075, -0.046662606, 1.3337431, -1.6876905, -0.40740314, -0.12567848, -0.74171, 0.33780485, 2.0305, -0.58057874, -1.112808, 0.90371263, -0.42485303, -1.6029017, 0.7744845, -0.54886556, 0.4346251, -0.18808684, -2.1350539, 0.2094973, 0.8783665, 0.41199994, 0.794365, 1.3964226, -0.16162448, 0.8851795, 1.7459514, -1.2020794, -0.75025433, 0.5759191, -1.9597256, 0.059071384, 0.94959474, 0.63463026, -0.8079282, 0.031265605, -0.8413961, -1.1042768, 0.6303067, -1.8922045, 0.44333157, -0.24039328, -1.1531141, 0.4702631, 1.3636328, 0.12676455, 1.1119264, 1.0851598, -0.17424987, 0.09117803, 0.5281344, -2.269342, 0.7240236, 1.2046725, -1.5445613, 0.43560714, 1.3055882, -0.08621345, 0.2435217, -0.45835873, 0.1720786, 0.7133352, 1.3131536, -1.8801373, -1.7852426, 0.683628, -0.06631989, 1.0313474, -0.47530708, -0.6839973, 0.09629285, 0.88116825, 0.9159164, 0.30998558, 1.8870829, -1.6558119, 0.27701497, -0.9132897, -1.6643144, 0.77425784, -0.114014894, -0.46283472, 0.4743833, 0.17162453, 0.3830775, -1.7981151, 1.4499048, -1.6079453, 1.0426321, -0.4928484, -0.70687276, -0.18113491, 0.69756544, -0.13894956, 1.2459149, 0.10677127, -1.2047523, -0.7041039, 0.748106, -0.77757025, -0.9965105, -0.56586826, -0.24453977, -0.19981901, 2.4897742, 0.44647083, -0.028593497, 1.0374067, 0.23086847, 0.1830584, 1.7442065, -0.14733131, -2.0698054, -0.04084452, -0.8674678, 1.4469824, 0.2182439, -0.5063364, -0.9308399, 0.73926544, -0.6499253, 0.63588357, 1.3024436, -0.9860947, -0.40290907, 0.58917207, -1.7343421, 0.30868053, 2.0506537, -0.6654691, -0.49161795, 0.04352488]`', /home/runner/work/burn/burn/burn-tensor/src/tensor/data.rs:231:13


failures:
    nn::transformer::decoder::tests::test_autoregressive_norm_last

test result: FAILED. 62 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.26s

Error: Process completed with exit code 101.

```

--- a/burn-tensor/src/tensor/data.rs
+++ b/burn-tensor/src/tensor/data.rs
@@ -3,7 +3,6 @@ use alloc::vec::Vec;
 
 use crate::{tensor::Shape, Element, ElementConversion};
 
-use libm::{pow, round};
 use rand::{distributions::Standard, Rng, RngCore};
 
 #[derive(serde::Serialize, serde::Deserialize, Debug, PartialEq, Eq, Clone, new)]
@@ -206,9 +205,14 @@ impl<E: core::fmt::Debug + Copy, const D: usize> Data<E, D> {
 
 impl<E: Into<f64> + Clone + core::fmt::Debug + PartialEq, const D: usize> Data<E, D> {
     pub fn assert_approx_eq(&self, other: &Self, precision: usize) {
-        assert_eq!(self.shape, other.shape);
-
-        let mut eq = true;
+        let mut message = String::new();
+        if self.shape != other.shape {
+            message += format!(
+                "\n  => Shape is different: {:?} != {:?}",
+                self.shape.dims, other.shape.dims
+            )
+            .as_str();
+        }
 
         let iter = self
             .value
@@ -216,19 +220,34 @@ impl<E: Into<f64> + Clone + core::fmt::Debug + PartialEq, const D: usize> Data<E
             .into_iter()
             .zip(other.value.clone().into_iter());
 
-        for (a, b) in iter {
+        let mut num_diff = 0;
+        let max_num_diff = 5;
+
+        for (i, (a, b)) in iter.enumerate() {
             let a: f64 = a.into();
             let b: f64 = b.into();
-            let a = round(pow(10.0_f64, precision as f64) * a);
-            let b = round(pow(10.0_f64, precision as f64) * b);
 
-            if a != b {
-                eq = false;
+            let err = libm::sqrt(libm::pow(a - b, 2.0));
+            let tolerance = libm::pow(0.1, precision as f64);
+
+            if err > tolerance {
+                // Only print the first 5 differents values.
+                if num_diff < max_num_diff {
+                    message += format!(
+                        "\n  => Position {i}: {a} != {b} | difference {err} > tolerance {tolerance}"
+                    )
+                    .as_str();
+                }
+                num_diff += 1;
             }
         }
 
-        if !eq {
-            assert_eq!(self.value, other.value);
+        if num_diff >= max_num_diff {
+            message += format!("\n{} more errors...", num_diff - 5).as_str();
+        }
+
+        if !message.is_empty() {
+            panic!("Tensors are not approx eq:{}", message);
         }
     }
 
@@ -386,5 +405,31 @@ mod tests {
 
         let data = Data::from([3.0, 5.0, 6.0]);
         assert_eq!(data.shape, Shape::new([3]));
+    }
+
+    #[test]
+    fn should_assert_appox_eq_limit() {
+        let data1 = Data::<f32, 2>::from([[3.0, 5.0, 6.0]]);
+        let data2 = Data::<f32, 2>::from([[3.01, 5.0, 6.0]]);
+
+        data1.assert_approx_eq(&data2, 2);
+    }
+
+    #[test]
+    #[should_panic]
+    fn should_assert_appox_eq_above_limit() {
+        let data1 = Data::<f32, 2>::from([[3.0, 5.0, 6.0]]);
+        let data2 = Data::<f32, 2>::from([[3.011, 5.0, 6.0]]);
+
+        data1.assert_approx_eq(&data2, 2);
+    }
+
+    #[test]
+    #[should_panic]
+    fn should_assert_appox_eq_check_shape() {
+        let data1 = Data::<f32, 2>::from([[3.0, 5.0, 6.0, 7.0]]);
+        let data2 = Data::<f32, 2>::from([[3.0, 5.0, 6.0]]);
+
+        data1.assert_approx_eq(&data2, 2);
     }
 }

--- a/burn-tensor/src/tensor/data.rs
+++ b/burn-tensor/src/tensor/data.rs
@@ -1,4 +1,5 @@
 use alloc::format;
+use alloc::string::String;
 use alloc::vec::Vec;
 
 use crate::{tensor::Shape, Element, ElementConversion};

